Forge Networking Remastered Unity/Assets/Bearded Man Studios Inc/Scripts/Networking/Forge/Networking/FacepunchP2PClient.cs

```diff
@@ -0,0 +1,453 @@
+
+			// By default pending creates should be true and flushed when ready
+			if (!pendCreates)
+				PendCreates = true;
+
+			try
+			{
+				ushort clientPort = DEFAULT_PORT;
+
+				// Make sure not to listen on the same port as the server for local networks
+				if (clientPort == DEFAULT_PORT)
+					clientPort++;
+
+				Client = new CachedFacepunchP2PClient(hostId);
+
+				// Do any generic initialization in result of the successful bind
+				OnBindSuccessful();
+
+				// Get a random hash key that needs to be used for validating that the server was connected to
+				headerHash = Websockets.HeaderHashKey();
+
+				// This is a typical Websockets accept header to be validated
+				byte[] connectHeader = Websockets.ConnectionHeader(headerHash, DEFAULT_PORT);
+
+				// Setup the identity of the server as a player
+				server = new NetworkingPlayer(0, hostId, true, this);
+
+				// Create the thread that will be listening for new data from connected clients and start its execution
+				Task.Queue(ReadNetwork);
+
+				//Let myself know I connected successfully
+				OnPlayerConnected(server);
+
+				// Set myself as a connected client
+				server.Connected = true;
+
+				//Set the port
+				SetPort(clientPort);
+
+				int connectCounter = 0;
+				Task.Queue(() =>
+				{
+					do
+					{
+						// Send the accept headers to the server to validate
+						Client.Send(connectHeader, connectHeader.Length, hostId, P2PSend.Reliable);
+						Thread.Sleep(3000);
+					} while (!headerExchanged && IsBound && ++connectCounter < CONNECT_TRIES);
+
+					if (connectCounter >= CONNECT_TRIES)
+					{
+						if (connectAttemptFailed != null)
+							connectAttemptFailed(this);
+					}
+				});
+
+				SteamNetworking.OnP2PConnectionFailed += OnP2PConnectionFailed;
+
+			}
+			catch (Exception e)
+			{
+				Logging.BMSLog.LogException(e);
+				// Do any generic initialization in result of the binding failure
+				OnBindFailure();
+
+				throw new FailedBindingException("Failed to bind to server's SteamId, see inner exception", e);
+			}
+		}
+
+		/// <summary>
+		/// Connects to a Steam Lobby
+		/// </summary>
+		/// <param name="lobbyToJoin">The <see cref="Steamworks.Data.Lobby"/> to join</param>
+		/// <param name="pendCreates">Set the NetWorker::PendCreates to true</param>
+		private async void ConnectToLobbyAsync(Steamworks.Data.Lobby lobbyToJoin, bool pendCreates)
+		{
+			await ConnectToLobby(lobbyToJoin, pendCreates);
+		}
+
+		/// <summary>
+		/// Enters a specified Steam Lobby and Connects to the lobby owner's SteamId asynchronously
+		/// </summary>
+		/// <param name="lobbyToJoin">The <see cref="Steamworks.Data.Lobby"/> to join</param>
+		/// <param name="pendCreates">Set the NetWorker::PendCreates to true</param>
+		private async System.Threading.Tasks.Task ConnectToLobby(Steamworks.Data.Lobby lobbyToJoin, bool pendCreates)
```

Comment:
Could `Steamworks.Data` be added as a `using` statement?

Reply:
there's a clash with the namespace BeardedManStudios.Forge.Networking and Steamworks.Data for Lobby, so I've consistently used Steamworks.Data to try to avoid ambiguity.

Reply:
okay